### ISSUE TITLE
SiteNav: loaded page not marked in blue if siteNav generated from an imported macro

### ISSUE THREAD

**Issue by damithc** [MarkBind/markbind#1261]
Consider this siteNav code:
```html
<navigation>

{% macro show_sitenav_items(items, is_flat=false) %}
* abc :expanded:
  * [xyz]({{ baseUrl }}/dg/index.html)
{% endmacro %}

{{ show_sitenav_items(dg_sitenav_items) }}

</navigation>
```
It results in this:
![image](https://user-images.githubusercontent.com/1673303/85870886-3cee9180-b800-11ea-945d-8bccf79065e6.png)

Note how `xyz` is marked in blue as that is the current page loaded.

Now, imagine the macro is moved to another page like this:
```html
<navigation>

{% from "scripts/macros.njk" import show_sitenav_items %}

{{ show_sitenav_items(dg_sitenav_items) }}

</navigation>
```

The site nav works as normal except the loaded page is not marked in blue anymore.
![image](https://user-images.githubusercontent.com/1673303/85871117-8b039500-b800-11ea-85f8-261b3266247c.png)

**Comment by ang-zeyu** [comment]
The link itself shouldn't work as well, this is by design of nunjucks. You could declare `baseUrl` as a parameter as a quick workaround though:

> Imported templates are processed without the current context, so they do not have access to any of the current template variables.

https://mozilla.github.io/nunjucks/templating.html#import

**Comment by damithc** [comment]
> The link itself shouldn't work as well, this is by design of nunjucks. You could declare `baseUrl` as a parameter as a quick workaround though:

Thanks for looking into it. That's not the problem though. The link works because `baseUrl` is a built-in global variable. I compared the HTML of the two files and it seems there is a subtle difference of a `&gt;` in one version.

Here is a live example of the problem https://reposense.org/dg/index.html

**Comment by ang-zeyu** [comment]
> The link works because baseUrl is a built-in global variable.

It works in this case because `baseUrl = ''`. Did a little more searching though, seems like nunjucks supports jinja's `{% from ... import ... with context %}`, although it is undocumented, could try that.

https://github.com/mozilla/nunjucks/issues/ 522
https://jinja.palletsprojects.com/en/2.9.x/templates/#import-context-behavior

> I compared the HTML of the two files and it seems there is a subtle difference of a &gt; in one version.

Which file / link is this?

**Comment by damithc** [comment]
> It works in this case because `baseUrl = ''`. Did a little more searching though, seems like nunjucks supports jinja's `{% from ... import ... with context %}`, although it is undocumented, could try that.

Good catch. Yes that seems to be the case. I usually throw in `with context` most of the time just in case but didn't use it for this particular case because I didn't think it matters for supposedly _global_ variables. This is the second time I got burned by leaving out `with context` (#1259 is the other). Thanks again for helping to solve this mystery. I've been trying to figure this out for a long time (happened in other places too).

> > I compared the HTML of the two files and it seems there is a subtle difference of a > in one version.
> 
> Which file / link is this?

Seems this was a false alarm. I guess the difference was caused by an event handler modifying the menu item to make it blue.

**Comment by ang-zeyu** [comment]
@damithc there is a quick solution around having to use `with context` for global variables, the downside being that it will change the priority of site variables to the lowest. (currently it has priority over page / imported variables)

should we go with that? I'm not sure of why site variables had higher priority than page /imported variables, since if you didn't want the page / imported variables, you could just omit them.

**Comment by damithc** [comment]
By _site variables_, do you mean things defined in the site.json?

**Comment by ang-zeyu** [comment]
nope, those in the respective (sub)site's `variables.md`

**Comment by damithc** [comment]
I think best to leave this alone for now. The `with context` works. I'd rather not go for a change that _might_ have unforeseen side effects, especially in the middle of the semester. We can revisit this later. Can reopen and put on hold if you prefer.

**Comment by ang-zeyu** [comment]
I think we can open this for V4, wdyt? @damithc 

The rationale again:

> should we go with that? I'm not sure of why site variables had higher priority than page /imported variables, since if you didn't want the page / imported variables, you could just omit them.

It's similar to the use case of `with context` or `<include><variable>...</variable></include>`, except we're overriding specific (page-wise) content with generic (site-wide) content, where it should be the opposite.

**Comment by damithc** [comment]
Sorry, I lost track of this.


> > should we go with that? I'm not sure of why site variables had higher priority than page /imported variables, since if you didn't want the page / imported variables, you could just omit them.
> 
> It's similar to the use case of `with context` or `<include><variable>...</variable></include>`, except we're overriding specific (page-wise) content with generic (site-wide) content, where it should be the opposite.

I think this relates to an important principle of text reuse which states that 'outer should be able to override the inner'.  This goes against the normal practice of inner overriding the outer which is what happens between a parent class and the sub class (the sub class can override the parent class). But it is the reverse in text reuse. That is, one should be able to take an existing piece of text and configure it to fit the reuse context (using configuration points in the text e.g., override the default values of variables), without modifying the reused code itself i.e., the _reuser_ overrides the _reusee_ As site.json sits at the top of the reuse chain, what's specified in there overrides what's stated in the page code.

**Comment by damithc** [comment]
> variables.md
> 
> * contains stuff that you would more than likely want to use _throughout_ your entire site (hence "site variables")

Yes.


> page X
> 
> * this page `<include> / {% include %}`'s some other source files referencing a site variable
> * an exception, that wants to use something else for the site variable. But, it would be extremely troublesome to change the source files, defeating content reuse.
> * I think ideally it should be able to do something like
>   ```
>   // in the case of {% include with context %}
>   {% set sitevariable = "new value..." %}
>   
>   // or in the case of <include>
>   <include zz><variable name="sitevariable">new value ....</variable</include>
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   then the included source files would also change

Yes, it's fine if the author can reuse another page (or a page fragment) while configuring it to fit the reuse context _at the point of reuse_.

Side note: `site.json` should be able to override frontmatter of a page though e.g., the `title` specified in the `site.json` should override the `title` in the frontmatter of the page.

**Comment by ang-zeyu** [comment]
_for a single site_, I think we established at least:

(higher the number higher the priority)
1. page variables (`{% set %}`s)
2. site variables (`variables.md`)
3. context variables (`<include zz><variable name="sitevariable">new value ....</variable</include>`)

(after re-reading the above) my original second point here actually has to do with `1` / `2`. I.e. they should be reversed.
_Within a single site_, the rationale is that you would never want to have a global configuration (variables.md) overriding individual page's values (`{% set %}`s). Because then it degenerates to not having defined the individual page's variable in the first place.
On the contrary you may want to use the same page variable name (as a site variable) to preserve some notion of variable semantics.

Does this make sense?
(before we go on to consider outer vs subsites)

> Side note: site.json should be able to override frontmatter of a page though e.g., the title specified in the site.json should override the title in the frontmatter of the page.

ps: this will still be preserved as they aren't variables.

**Comment by damithc** [comment]
> generates: `filename is variables.md` still.
> 
> The change would ideally allow this override to happen: `filename is page5.md`

Thanks for the clarification @ang-zeyu Sounds good 👍
